### PR TITLE
Leaderboards of finished games

### PR DESCRIPTION
**Pull request by Eren361**
I wanted this to enable users to view leaderboard of a lobby.
I updated lobby, repository and routes. Added a html ui. Added a GetLeaderboard method, created HTTP handler and another method in lobby to return top 3 players.

### CODE REVIEW

File: internal/lobby/routes.go [hunk 1]
```diff
@@ -29,6 +29,7 @@ func NewLobbiesRouter() http.Handler {
 	mux.HandleFunc("GET /lobbies/{$}", getLobbiesHandler)
 	mux.HandleFunc("GET /lobbies/{pin}", getLobbyByPinHandler)
 	mux.HandleFunc("/lobbies/{pin}/ws", getLobbyByPinWsHandler)
+	mux.HandleFunc("/lobbies/leaderboard", getLeaderboardHandler)
```
Comment (erykksc) [inline internal/lobby/routes.go:32]: Move this handler to package "results" (you need to create it) and change the path to "/results/{pin}"

File: internal/lobby/repository.go
Comment (erykksc) [inline internal/lobby/repository.go:1]: Lobbies are active games, they are not permanent.
The results of the game will be saved inside the common.Game struct located in "internal/common/types".
So move/adapt the methods from Lobby to Game.

File: internal/lobby/lobby.go [hunk 2]
```diff
@@ -48,6 +48,7 @@ func NewClientID() (ClientID, error) {
 type Player struct {
 	Conn     *websocket.Conn
 	Username string
+	Score 	 int
```
Comment (erykksc) [inline internal/lobby/lobby.go:51]: The score for a player is gonna be saved inside "Points" field on common.Game struct located in "internal/common/types".

File: internal/lobby/routes.go [hunk 3]
```diff
@@ -49,6 +50,34 @@ func NewLobbiesRouter() http.Handler {
...
+
+	leaderboard, err := lobbiesRepo.GetLeaderboard(pin)
+	if err != nil {
+		if _, ok := err.(ErrLobbyNotFound); ok {
+			common.ErrorHandler(w, r, http.StatusNotFound)
+			return
+		}
+		common.ErrorHandler(w, r, http.StatusInternalServerError)
+		return
+	}
```
Comment (erykksc) [inline internal/lobby/routes.go:70]: As stated in other comments, the lobbies are only temporary, implement the Methods like GetLeaderboard on common.Game struct

File: internal/lobby/lobby.go [hunk 4]
```diff
@@ -64,3 +65,23 @@ func CreateLobby(options LobbyOptions) *Lobby {
...
+
+	sort.Slice(players, func(i, j int) bool {
+		return players[i].Score > players[j].Score
+	})
+
+	if len(players) > 3 {
+		return players[:3]
+	}
+	return players
+}
```
Comment (erykksc) [inline internal/lobby/lobby.go:87]: Lobbies are active games, they are not permanent.
The results of the game will be saved inside the common.Game struct located in "internal/common/types".
So move/adapt the methods from Lobby to Game.

**Review by erykksc — CHANGES_REQUESTED** [review]
Generally move the implementation of the past games into "results" package (you need to create it)

**Comment by Eren361** [comment]
> It doesn't compile. Please fix it.

go run "/Users/erenkayakiran/Studium/Praktikum/kwikquiz/cmd/kwikquiz/main.go"
(base) erenkayakiran@Erens-MacBook-Air kwikquiz % go run "/Users/erenkayakiran/Studium/Praktikum/kwikquiz/cmd/kwikquiz/main.go"
time=2024-06-06T00:20:45.204+02:00 level=INFO msg="Server listening" addr=:3000

it compiles by me, can you give more feedback?

**Pull request closed without merging**